python/pyarrow/array.pxi

```diff
@@ -841,6 +841,223 @@ cdef class Array(_PandasConvertible):
         return res
 
 
+cdef class SparseTensorCOO:
+    """
+    A sparse COO tensor.
+    """
+
+    def __init__(self):
+        raise TypeError("Do not call SparseTensorCOO's constructor directly, "
+                        "use one of the `pyarrow.SparseTensorCOO.from_*` "
+                        "functions instead.")
+
+    cdef void init(self, const shared_ptr[CSparseTensorCOO]& sp_sparse_tensor):
+        self.sp_sparse_tensor = sp_sparse_tensor
+        self.stp = sp_sparse_tensor.get()
+        self.type = pyarrow_wrap_data_type(self.stp.type())
+
+    def __repr__(self):
+        return """<pyarrow.SparseTensorCOO>
+type: {0.type}
+shape: {0.shape}""".format(self)
+
+    @staticmethod
+    def from_dense_numpy(obj):
+        cdef shared_ptr[CSparseTensorCOO] csparse_tensor
+        with nogil:
+            check_status(NdarrayToSparseTensorCOO(c_default_memory_pool(), obj,
+                                                  &csparse_tensor))
+        return pyarrow_wrap_sparse_tensor_coo(csparse_tensor)
+
+    @staticmethod
+    def from_numpy(data, coords, shape, dim_names=None):
+        cdef shared_ptr[CSparseTensorCOO] csparse_tensor
+        cdef vector[int64_t] c_shape
+        cdef vector[c_string] c_dim_names
+
+        for x in shape:
+            c_shape.push_back(x)
+        if dim_names is not None:
+            for x in dim_names:
+                c_dim_names.push_back(tobytes(x))
+        with nogil:
+            check_status(NdarraysToSparseTensorCOO(c_default_memory_pool(),
+                         data, coords, c_shape, c_dim_names, &csparse_tensor))
+        return pyarrow_wrap_sparse_tensor_coo(csparse_tensor)
+
+    @staticmethod
+    def from_tensor(obj):
+        cdef shared_ptr[CSparseTensorCOO] csparse_tensor
+        cdef shared_ptr[CTensor] ctensor = pyarrow_unwrap_tensor(obj)
+
+        with nogil:
+            check_status(TensorToSparseTensorCOO(ctensor, &csparse_tensor))
+
+        return pyarrow_wrap_sparse_tensor_coo(csparse_tensor)
+
+    def to_numpy(self):
+        """
+        Convert arrow::SparseTensorCOO to numpy.ndarrays with zero copy
+        """
+        cdef PyObject* out_data
+        cdef PyObject* out_coords
+
+        with nogil:
```

Comment:
It doesn't make sense to use `nogil` since the C++ function below takes the GIL.

Reply:
Got it. Removing.